src/mapport.cpp

```diff
@@ -46,104 +43,96 @@ using namespace std::chrono_literals;
+
+            // Open a port mapping on whatever local address we have toward the gateway.
+            struct in_addr inaddr_any;
+            inaddr_any.s_addr = htonl(INADDR_ANY);
+            auto res = PCPRequestPortMap(pcp_nonce, *gateway4, CNetAddr(inaddr_any), private_port, requested_lifetime);
+            MappingError* pcp_err = std::get_if<MappingError>(&res);
+            if (pcp_err && *pcp_err == MappingError::UNSUPP_VERSION) {
+                LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "portmap: Got unsupported PCP version response, falling back to NAT-PMP\n");
+                res = NATPMPRequestPortMap(*gateway4, private_port, requested_lifetime);
+            }
+            handle_mapping(res);
         }
-    } else {
-        LogPrintf("natpmp: sendpublicaddressrequest() failed with %d error.\n", r_send);
-    }
 
-    return false;
-}
+        // IPv6
+        std::optional<CNetAddr> gateway6 = QueryDefaultGateway(NET_IPV6);
+        if (!gateway6) {
+            LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "portmap: Could not determine IPv6 default gateway\n");
+        } else {
+            LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "portmap: gateway [IPv6]: %s\n", gateway6->ToStringAddr());
 
-static bool NatpmpMapping(natpmp_t* natpmp, const struct in_addr& external_ipv4_addr, uint16_t private_port, bool& external_ip_discovered)
-{
-    const uint16_t suggested_external_port = g_mapport_external_port ? g_mapport_external_port : private_port;
-    const int r_send = sendnewportmappingrequest(natpmp, NATPMP_PROTOCOL_TCP, private_port, suggested_external_port, 3600 /*seconds*/);
-    if (r_send == 12 /* OK */) {
-        int r_read;
-        natpmpresp_t response;
-        do {
-            r_read = readnatpmpresponseorretry(natpmp, &response);
-        } while (r_read == NATPMP_TRYAGAIN);
-
-        if (r_read == 0) {
-            auto pm = response.pnu.newportmapping;
-            if (private_port == pm.privateport && pm.lifetime > 0) {
-                g_mapport_external_port = pm.mappedpublicport;
-                const CService external{external_ipv4_addr, pm.mappedpublicport};
-                if (!external_ip_discovered && fDiscover) {
-                    AddLocal(external, LOCAL_MAPPED);
-                    external_ip_discovered = true;
-                }
-                LogPrintf("natpmp: Port mapping successful. External address = %s\n", external.ToStringAddrPort());
-                return true;
-            } else {
-                LogPrintf("natpmp: Port mapping failed.\n");
+            // Try to open pinholes for all routable local IPv6 addresses.
+            for (const auto &addr: GetLocalAddresses()) {
+                if (!addr.IsRoutable() || !addr.IsIPv6()) continue;
+                auto res = PCPRequestPortMap(pcp_nonce, *gateway6, addr, private_port, requested_lifetime);
```

Comment:
Also, it just occurred to me, `GetLocalAddresses()` operates regardless of `-bind=` options. If the machine has two IPv6 addresses but `-bind=` has been used and only one of them specified, then we shouldn't do mapping for the other one (where nobody is listening).

Reply:
>  Why make mappings for all local IPv6 addresses?

It's the sensible choice IMO. Some routers will assign a whole bunch of routable IPv6 addresses but only allow pinholing for some. It's doesn't hurt to try.

> Also, it just occurred to me, GetLocalAddresses() operates regardless of -bind= option

Yes. But mind that `-natpmp` and `-pnp` are set to 0 if  `-bind` is specified. In that case we assume the user has a more complex configuration and will just do their own mapping.

Reply:
>> I think it is going to result in multiple mappings on the same external IPv6 address but on different ports.

> i don't think so? We only request one mapping per IPv6 address at most.

Afaik PCP always uses a pinhole, which does not involve assinging a (different) port.

We do end up with our node being announced multiple times with different IP addresses, which could cause some other node to connect to use twice. That can already happen when you have IPv4 and IPv6 (and Tor and I2P) enabled.

If we're really worried about this, it might make sense to try each IPv6 address sequentially and stop when one is acknowledged. I think that can wait for a followup.

Reply:
Ok, I was having NAT in mind when I wrote "is going to result in multiple mappings on the same external IPv6 address". Sorry for the noise.

Reply:
In addition to the `-bind` [thing](https://github.com/bitcoin/bitcoin/pull/30043#discussion_r1658949236) there is also a complication with the port: if `-bind=10.0.0.1:20000 -bind=[5566::1]:30000` has been used, then `GetListenPort()`, currently used by `mapport.cpp`, is not the right thing to do because it would return `20000`.

To resolve both, we could get the listening addresses + the correct ports from:
1. `AddLocal()` / `mapLocalHost`, is called after bind (good), but also from other places and after the bind it is called only if `fDiscover` is true (bad).
2. `CConnman::vhListenSocket`, has sockets that used `BF_DONT_ADVERTISE` during bind. Those should be omitted.
3. `-bind` arguments, could get it from `init.cpp / connOptions`. Need to duplicate some of the logic of `CConnman::InitBinds()`.
4. introduce a new global list of `addr:port`, similar but not identical to `mapLocalHost`, seems imperfect because the list is almost the same as `mapLocalHost`, ideally those addresses should be in one place with proper flags to distinguish them (or maybe not?).

Here is a crude implementation of 4:

<details>
<summary>[patch] do 4.</summary>

```diff
diff --git a/src/mapport.cpp b/src/mapport.cpp
index a1d584dae9..56d21e792b 100644
--- a/src/mapport.cpp
+++ b/src/mapport.cpp
@@ -48,13 +48,13 @@ static bool ProcessPCP()
     // The same nonce is used for all mappings, this is allowed by the spec, and simplifies keeping track of them.
     PCPMappingNonce pcp_nonce;
     GetRandBytes(pcp_nonce);
 
     bool ret = false;
     bool no_resources = false;
-    const uint16_t private_port = GetListenPort();
+    const uint16_t private_port = GetListenPort(); // XXX could be the wrong port if -bind is used
     // Multiply the reannounce period by two, as we'll try to renew approximately halfway.
     const uint32_t requested_lifetime = std::chrono::seconds(PORT_MAPPING_REANNOUNCE_PERIOD * 2).count();
     uint32_t actual_lifetime = 0;
     std::chrono::milliseconds sleep_time;
 
     // Local functor to handle result from PCP/NATPMP mapping.
@@ -101,14 +101,18 @@ static bool ProcessPCP()
         if (!gateway6) {
             LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "portmap: Could not determine IPv6 default gateway\n");
         } else {
             LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "portmap: gateway [IPv6]: %s\n", gateway6->ToStringAddr());
 
             // Try to open pinholes for all routable local IPv6 addresses.
-            for (const auto &addr: GetLocalAddresses()) {
-                if (!addr.IsRoutable() || !addr.IsIPv6()) continue;
+            const auto addresses =
+                WITH_LOCK(g_addresses_for_port_map.mutex, return g_addresses_for_port_map.addresses);
+            for (const auto& addr : addresses) {
+                if (!addr.IsIPv6()) {
+                    continue;
+                }
                 auto res = PCPRequestPortMap(pcp_nonce, *gateway6, addr, private_port, requested_lifetime);
                 handle_mapping(res);
             }
         }
 
         // Log message if we got NO_RESOURCES.
diff --git a/src/net.cpp b/src/net.cpp
index 7c6babc389..b621e76696 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -112,12 +112,15 @@ static const uint64_t RANDOMIZER_ID_ADDRCACHE = 0x1cf2e4ddd306dda9ULL; // SHA256
 // Global state variables
 //
 bool fDiscover = true;
 bool fListen = true;
 GlobalMutex g_maplocalhost_mutex;
 std::map<CNetAddr, LocalServiceInfo> mapLocalHost GUARDED_BY(g_maplocalhost_mutex);
+
+AddressesForPortMap g_addresses_for_port_map;
+
 std::string strSubVersion;
 
 size_t CSerializedNetMsg::GetMemoryUsage() const noexcept
 {
     // Don't count the dynamic memory used for the m_type string, by assuming it fits in the
     // "small string" optimization area (which stores data inside the object itself, up to some
@@ -3151,14 +3154,17 @@ bool CConnman::Bind(const CService& addr_, unsigned int flags, NetPermissionFlag
         if ((flags & BF_REPORT_ERROR) && m_client_interface) {
             m_client_interface->ThreadSafeMessageBox(strError, "", CClientUIInterface::MSG_ERROR);
         }
         return false;
     }
 
-    if (addr.IsRoutable() && fDiscover && !(flags & BF_DONT_ADVERTISE) && !NetPermissions::HasFlag(permissions, NetPermissionFlags::NoBan)) {
-        AddLocal(addr, LOCAL_BIND);
+    if (addr.IsRoutable() && !(flags & BF_DONT_ADVERTISE) && !NetPermissions::HasFlag(permissions, NetPermissionFlags::NoBan)) {
+        if (fDiscover) {
+            AddLocal(addr, LOCAL_BIND);
+        }
+        WITH_LOCK(g_addresses_for_port_map.mutex, g_addresses_for_port_map.addresses.push_back(addr));
     }
 
     return true;
 }
 
 bool CConnman::InitBinds(const Options& options)
diff --git a/src/net.h b/src/net.h
index 11cb01a95d..ae08603e64 100644
--- a/src/net.h
+++ b/src/net.h
@@ -175,12 +175,17 @@ struct LocalServiceInfo {
     uint16_t nPort;
 };
 
 extern GlobalMutex g_maplocalhost_mutex;
 extern std::map<CNetAddr, LocalServiceInfo> mapLocalHost GUARDED_BY(g_maplocalhost_mutex);
 
+extern struct AddressesForPortMap {
+    Mutex mutex;
+    std::vector<CService> addresses GUARDED_BY(mutex);
+} g_addresses_for_port_map;
+
 extern const std::string NET_MESSAGE_TYPE_OTHER;
 using mapMsgTypeSize = std::map</* message type */ std::string, /* total bytes */ uint64_t>;
 
 class CNodeStats
 {
 public:
```
</details>